tower-http/src/follow_redirect/mod.rs

```diff
@@ -0,0 +1,325 @@
+                *req.headers_mut() = this.headers.clone();
+                this.policy.on_request(&mut req);
+                this.future.set(this.service.call(req));
+
+                cx.waker().wake_by_ref();
+                Poll::Pending
+            }
+            ActionKind::Stop => Poll::Ready(Ok(res)),
+            ActionKind::Error(e) => Poll::Ready(Err(e)),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum BodyRepr<B> {
+    Some(B),
+    Empty,
+    None,
+}
+
+impl<B> BodyRepr<B>
+where
+    B: Body + Default,
+{
+    fn take(&mut self) -> Option<B> {
+        match mem::replace(self, BodyRepr::None) {
+            BodyRepr::Some(body) => Some(body),
+            BodyRepr::Empty => {
+                *self = BodyRepr::Empty;
+                Some(B::default())
+            }
+            BodyRepr::None => None,
+        }
+    }
+
+    fn try_clone_from<P, E>(&mut self, body: &B, policy: &P)
+    where
+        P: Policy<B, E>,
+    {
+        match self {
+            BodyRepr::Some(_) | BodyRepr::Empty => {}
+            BodyRepr::None => {
+                if let Some(body) = clone_body(policy, body) {
+                    *self = BodyRepr::Some(body);
+                }
+            }
+        }
+    }
+}
+
+fn clone_body<P, B, E>(policy: &P, body: &B) -> Option<B>
+where
+    P: Policy<B, E>,
+    B: Body + Default,
+{
+    if body.size_hint().exact() == Some(0) {
+        Some(B::default())
+    } else {
+        policy.clone_body(body)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{policy::*, *};
+    use hyper::{header::LOCATION, Body};
+    use std::convert::Infallible;
+    use tower::{ServiceBuilder, ServiceExt};
+
+    #[tokio::test]
+    async fn follows() {
+        let svc = ServiceBuilder::new()
+            .layer(FollowRedirectLayer::new(Action::follow()))
+            .service_fn(handle);
+        let req = Request::builder()
+            .uri("http://example.com/42")
+            .body(Body::empty())
+            .unwrap();
+        let res = svc.oneshot(req).await.unwrap();
+        assert_eq!(res.into_body(), 0);
+    }
+
+    #[tokio::test]
+    async fn stops() {
+        let svc = ServiceBuilder::new()
+            .layer(FollowRedirectLayer::new(Action::stop()))
+            .service_fn(handle);
+        let req = Request::builder()
+            .uri("http://example.com/42")
+            .body(Body::empty())
+            .unwrap();
+        let res = svc.oneshot(req).await.unwrap();
+        assert_eq!(res.into_body(), 42);
+    }
+
+    #[tokio::test]
+    async fn limited() {
+        let svc = ServiceBuilder::new()
+            .layer(FollowRedirectLayer::new(Limited::new(10)))
+            .service_fn(handle);
+        let req = Request::builder()
+            .uri("http://example.com/42")
+            .body(Body::empty())
+            .unwrap();
+        let res = svc.oneshot(req).await.unwrap();
+        assert_eq!(res.into_body(), 42 - 10);
+    }
```

Comment:
I'm thinking a test where the inner service is wrapper in a `tower::buffer::Buffer` and does a couple of redirects would be cool. `Buffer` should highlight any mistakes in how calls to `poll_ready`/`call` are managed.

Reply:
Added in 35e66afc40078576fe27072e3f58742b3588889a. It seems to be able to catch mistakes in both `FollowRedirect::call` and `ResponseFuture::poll`.